activation/activation.go

```diff
@@ -243,50 +250,82 @@ func (b *Builder) buildNipstChallenge(epoch types.EpochId) error {
 }
 
 func (b *Builder) StartPost(rewardAddress address.Address, dataDir string, space uint64) error {
+	b.InitLock.Lock()
+	switch b.initStatus {
+	case InitDone:
+		b.InitLock.Unlock()
+		return fmt.Errorf("already initialized")
+	case InitInProgress:
+		b.InitLock.Unlock()
+		return fmt.Errorf("already started")
+	}
+
 	b.log.Info("Starting post, reward address: %x", rewardAddress)
+
+	b.initStatus = InitInProgress
+	b.postProver.SetParams(dataDir, space)
+	b.InitLock.Unlock()
+
 	b.SetCoinbaseAccount(rewardAddress)
-	b.postInitLock.Lock()
-	if b.initStatus == Done {
-		b.postInitLock.Unlock()
-		return nil
+
+	initialized, err := b.postProver.IsInitialized()
+	if err != nil {
+		b.InitLock.Lock()
+		b.initStatus = InitIdle
+		b.InitLock.Unlock()
+		return err
 	}
-	if b.initStatus == InProgress {
-		b.postInitLock.Unlock()
-		return fmt.Errorf("attempted to start post when post already initiated")
+
+	if !initialized {
+		if err := b.postProver.VerifyInitAllowed(); err != nil {
+			b.InitLock.Lock()
+			b.initStatus = InitIdle
+			b.InitLock.Unlock()
+			return err
+		}
 	}
-	b.initStatus = InProgress
-	errChan := make(chan error, 1)
+
 	go func() {
-		_, err := b.nipstBuilder.InitializePost(dataDir, space) // TODO: add proof to first ATX
-		b.postInitLock.Lock()
-		if err != nil {
-			b.initStatus = Idle
-			b.postInitLock.Unlock()
-			b.log.Error(err.Error())
-			errChan <- err
-			return
+		if initialized {
+			b.commitment, err = b.postProver.Execute(shared.ZeroChallenge)
+			if err != nil {
+				b.log.Error("PoST execution failed: %v", err)
+
+				b.InitLock.Lock()
+				b.initStatus = InitIdle
+				b.InitLock.Unlock()
+				return
+			}
+		} else {
+			b.commitment, err = b.postProver.Initialize()
+			if err != nil {
+				b.log.Error("PoST initialization failed: %v", err)
+
+				b.InitLock.Lock()
+				b.initStatus = InitIdle
+				b.InitLock.Unlock()
+				return
+			}
 		}
```

Comment:
I'm not sure on what code did you wrote that comment.
But the initial post _is_ being run in case that `initialized` is false.
The only case we're not doing that is when `VerifyInitAllowed` returns an err.
We're doing this check before launching the goroutine, to return an immediate err to the client.